### PR TITLE
First pass at working test runner

### PR DESCRIPTION
**Pull request by joshgoebel**
Could use thoughts/feedback.  No need to pay a ton of attention to `vendor`, I still need to clean it up further.  Essentially we're replacing the `vendor` from the wren track (which includes critical testing libs, etc) with our own version suited for test-runner usage.

- Follows the template repository structure overall
- `vendor` is symlinked into each `test/**/input` directory prior to the tests being run
- Adds `tests/**/data.json` to allow per-test metadata, such as the `slug` which we use to find the test file - is there some better way (or established pattern) to do this?
- `run-tests.sh` is pretty boring other than the use of data.json for metadata
- `run.sh` pretty much calls `wren_cli [test_file] [result_file]`
- A custom subclass of Testie provides a `JSONReporter` which dumps the test results into JSON instead of presenting them visually.
- Only the first error generates a full stack trace (we have to re-run the first failing test without guardrails and crash the test runner in order to get this).  The JSON output is written prior to crashing. 
- The stack trace output is detected with some `sed` magic and then merged back into the already written JSON. 
- If the test runner itself blows up unexpectedly (syntax error, etc) then we fall back to `status:error` with the full output

Docker:

- Using alpine base, two stage build.
- Compiling `wren_cli` from Github with my file extension patch

Questions:

- Should we use the slug to find the tests or just find the first `*.spec.wren` file and just assume that's good enough?
- Docker breaks because it's being run `--read-only` yet the template repo has the tests writing into their own folders, so I'm at a bit of a loss...?  We really need the filesystem writable to symlink `vendor` also...

### CODE REVIEW

**Comment by joshgoebel** [comment]
@ErikSchierboom Could use your high-level thoughts when you have a chance.

**Comment by ErikSchierboom** [comment]
> Should we use the slug to find the tests or just find the first *.spec.wren file and just assume that's good enough?

You could do either, both would work. The nicest way to do it though is by reading the `files.test` key from the exercise's `.meta/config.json` file (see https://github.com/exercism/docs/blob/main/building/tracks/practice-exercises.md#file-metaconfigjson).

> Docker breaks because it's being run --read-only yet the template repo has the tests writing into their own folders, so I'm at a bit of a loss...?

The idea is that the entire file system is read-only, except for the passed-in input and output directories, as well as the /tmp directory. This helps reduce the attack surface for any attacker.

> We really need the filesystem writable to symlink vendor also...

Other test runners "solve" this by just copying the source files to the input directory, which is usually more than fast enough for our purposes.

File: bin/run.sh
Comment (ErikSchierboom) [inline bin/run.sh:51]: What does `jq -M` do? I don't see any query.

File: Dockerfile
Comment (ErikSchierboom) [inline Dockerfile:4]: Do you need `jq` and/or `bash` for this build stage?

**Review by ErikSchierboom — APPROVED** [review]
LGTM. Some minor comments

File: bin/run-tests.sh
Comment (joshgoebel) [reply bin/run-tests.sh:34]: Provides the standard diff headers that I consider "canonical"... I've never use diff any other way.

**Comment by joshgoebel** [comment]
> The nicest way to do it though is by reading the files.test key from the exercise's .meta/config.json file (see https://github.com/exercism/docs/blob/main/building/tracks/practice-exercises.md#file-metaconfigjson).

So will that whole directory be there magically in `input` along with the students solution? 

> The idea is that the entire file system is read-only ... helps reduce the attack surface for any attacker.

Yeah I understood that.  It's sorted now. :)  I thought `input` was also read-only but I wasn't looking carefully.

> Other test runners "solve" this by just copying the source files to the input directory, which is usually more than fast enough for our purposes.

This is all happening inside `input`, where the users code is... symlinking seems to work just fine.  The only issue was my trying to put a temp file in whatever the working directory was, which is a no go.

File: bin/run-tests.sh
Comment (joshgoebel) [reply bin/run-tests.sh:35]: If this proves a problem for other maintainers it can change in the future.  Personally I'd prefer not to have "build assets" laying around if I can avoid it.  Perhaps you should try `diff -urN`. 😇

**Comment by ErikSchierboom** [comment]
> So will that whole directory be there magically in input along with the students solution?

Correct.

File: bin/run.sh
Comment (joshgoebel) [reply bin/run.sh]: monochrome, but looks like it's not necessary.

File: Dockerfile
Comment (joshgoebel) [reply Dockerfile:4]: Nope, i've since removed.

**Comment by joshgoebel** [comment]
This still bothers me:

```
if [ -f "${test_dir_path}/data.json" ]; then
        slug=$(jq -r .slug ${test_dir_path}/data.json)
    fi
```

It's making me realize it's quite strange that `slug` isn't part of `meta/config.json`... There should be a way to find an exercises slug that ISN'T dependent on it's source folder - since in some cases (like here) we don't have the source folder...

**Comment by iHiD** [comment]
> It's making me realize it's quite strange that slug isn't part of meta/config.json... There should be a way to find an exercises slug that ISN'T dependent on it's source folder - since in some cases (like here) we don't have the source folder...

https://github.com/exercism/wren-test-runner/blob/main/bin/run.sh#L7

**Comment by joshgoebel** [comment]
> https://github.com/exercism/wren-test-runner/blob/main/bin/run.sh#L7

I'm talking about the test suite of the test runner... yes, in *production* the slug is passed.. but I'm trying to run all my test runner tests locally (for CI, etc) and EVEN if I copy the ENTIRE exercise from the track into the test runner, that metadata [slug] is STILL not present. (unless you make assumptions about the file names).  _I need the slug so I can pass it to the test runner to properly simulate production._

It feels like an omission that `config.json` for an exercise doesn't have the slug.  Having exercise meta-data 'hidden' in the directory structure seems suboptimal.  So because of this lack I have to maintain my own `data.json` files for every single test just to contain:

```
{
slug: "two-fer"
}
```

That was my question.  Why can't this be moved into `config.json` for the exercises?

File: bin/run.sh
Comment (joshgoebel) [inline bin/run.sh:67]: Is there some magic to printf vs echo here? @ErikSchierboom

Comment (joshgoebel) [reply bin/run.sh:67]: I forget why I changed it.

**Comment by iHiD** [comment]
@joshgoebel You can add your exercise slugs to the `config.json` files. You can put any extra fields in there. 

In terms of whether this should be part of the wider standard, I can see a good argument for that. @ErikSchierboom?

File: bin/run.sh
Comment (ErikSchierboom) [reply bin/run.sh:67]: I can't even remember :D

**Comment by ErikSchierboom** [comment]
> unless you make assumptions about the file names

This is exactly what tracks do: `configlet` verifies that the exercise's directory exactly matches the slug as specified in the `config.json`, which means that relying on the directory name _can_ be done safely. 

> @joshgoebel You can add your exercise slugs to the config.json files. You can put any extra fields in there.

This is exactly what I would propose. Drop the `data.json` files and merge its contents into the `.meta/config.json` files. You can add any additional fields to `.meta/config.json` files.

> In terms of whether this should be part of the wider standard, I can see a good argument for that. @ErikSchierboom?

I don't know, I've not heard of other tracks requesting this feature. The track `config.json` currently contains the slug, which makes a lot of sense as otherwise maintainers wouldn't be able to easily identify the exercise just based on its `uuid`. I have no objections to the slug being duplicated in the `.meta/config.json` file, but I would argue that individual tracks should decide if they want that themselves, I'd prefer not to have this as a wider standard.

**Pull request merged**
